printer.ml

```diff
@@ -160,18 +160,36 @@ let name_of tm =
 (* Printer for types.                                                        *)
 (* ------------------------------------------------------------------------- *)
 
-let rec sp_print_type fmt ty =
-  let rec sp_print_type_list tyl =
-    match tyl with
-      [] -> ()
-    | t::tl -> pp_print_string fmt " ";
-               sp_print_type fmt t;
-               sp_print_type_list tl
-  in match ty with
-       Tyvar(str) -> pp_print_string fmt str
-     | Tyapp(str,tyl) -> pp_print_string fmt ("("^str);
-                         sp_print_type_list tyl;
-                         pp_print_string fmt ")"
+type sexp =
+  | Sleaf of string
+  | Snode of sexp list
+
+let rec sexp_print fmt s = match s with
+    Sleaf s -> pp_print_string fmt s
+  | Snode [] -> pp_print_string fmt "()"
+  | Snode (s::sl) ->
+      pp_print_char fmt '(';
+      sexp_print fmt s;
+      List.iter (fun s -> pp_print_char fmt ' '; sexp_print fmt s) sl;
+      pp_print_char fmt ')';;
+
+(* Turn a converter raw : 'a -> sexp into a memoized converter memo : 'a -> sexp
+   that writes either (memo _n raw) or _n, where _n is fresh. *)
+let sexp_memoize : ('a -> sexp) -> ('a -> sexp) =
+  let next = ref 1 in
+  fun raw ->
+    let memo = Hashtbl.create 100 in
```

Comment:
Leave a todo to update this number once we have better stats on the average number of unique expressions are in each proof_log.